src/inmanta/execute/dataflow/datatrace.py

```diff
@@ -0,0 +1,241 @@
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+from itertools import chain
+from typing import TYPE_CHECKING, Iterable, List, Optional
+
+from inmanta.ast import Locatable, NotFoundException
+from inmanta.ast.statements import Statement
+from inmanta.execute.dataflow import (
+    AssignableNode,
+    AssignableNodeReference,
+    Assignment,
+    AttributeNode,
+    AttributeNodeReference,
+    DataflowGraph,
+    Equivalence,
+    InstanceNode,
+    InstanceNodeReference,
+    VariableNodeReference,
+)
+from inmanta.execute.runtime import Instance
+
+if TYPE_CHECKING:
+    pass
+
+
+class DataTraceRenderer:
+    @classmethod
+    def _prefix_line(cls, prefix: str, line: str) -> str:
+        """
+            Prefixes a line.
+        """
+        if line == "":
+            return prefix.rstrip()
+        return prefix + line
+
+    @classmethod
+    def _prefix(cls, prefix: str, lines: Iterable[str]) -> List[str]:
+        """
+            Prefixes lines.
+        """
+        return [cls._prefix_line(prefix, line) for line in lines]
+
+    @classmethod
+    def _branch(cls, lines: List[str], last: Optional[bool] = False) -> List[str]:
+        """
+            Renders a branch in the tree.
+
+            :param last: True iff this is the last branch on this level.
+        """
+        if len(lines) == 0:
+            return []
+        branch_prefix: str = ("└" if last else "├") + "── "
+        block_prefix: str = (" " if last else "│") + " " * 3
+        result: List[str] = cls._prefix(branch_prefix, lines[0:1])
+        result += cls._prefix(block_prefix, lines[1:])
+        return result
+
+    @classmethod
+    def _indent(cls, lines: Iterable[str]) -> List[str]:
+        """
+            Indents lines.
+        """
+        return cls._prefix(" " * 4, lines)
+
+    @classmethod
+    def _render_implementation_context(cls, context: DataflowGraph) -> List[str]:
+        """
+            Renders information about the dynamic implementation context, if it exists.
+        """
+        try:
+            context.resolver.lookup("self")
+        except NotFoundException:
+            return []
+        result: List[str] = []
+        var_node: AssignableNodeReference = context.resolver.get_dataflow_node("self")
+        if (
+            isinstance(var_node, VariableNodeReference)
+            and len(var_node.node.instance_assignments) == 1
+            and isinstance(var_node.node.instance_assignments[0].responsible, Instance)
+        ):
+            instance_node: InstanceNode = var_node.node.instance_assignments[0].rhs.top_node()
+            result.append("IN IMPLEMENTATION WITH self = %s" % instance_node)
```

Comment:
It is at this point not specific which implementation or even how many. This is printed elsewhere?

Reply:
Actually no, I didn't think of that. I'll add that as well.